qemu/tests/balloon_boot_in_pause.py

```diff
@@ -0,0 +1,221 @@
+    """
+
+    def get_memory_status(self):
+        """
+        Get Memory status inside guest
+
+        :return: the used memory size inside guest.
+        :rtype: int
+        """
+        return int(self.get_used_mem())
+
+
+class BallooningTestPauseLinux(BallooningTestPause):
+
+    """
+    Linux memory ballooning test for guest booted in paused status
+    """
+
+    def get_memory_status(self):
+        """
+        Get Memory status inside guest
+
+        :return: the total memory size inside guest.
+        :rtype: int
+        """
+        return int(self.get_total_mem())
+
+
+@error_context.context_aware
+def run(test, params, env):
+    """
+    Balloon guest memory when guest started in paused status,
+    use M when compare memory in this script:
+    1) Boot a guest with balloon enabled and in paused status,
+    i.e. '-S' used but not cont
+    2) Evict guest memory in paused status, cont the guest;
+    check memory in monitor
+    3) To check if the guest memory balloon working well after above test,
+    continue to do:
+    3.1) Enlarge guest memory in running status;
+    check memory both in guest and monitor
+    3.2) Evict guest memory in running status;
+    check memory both in guest and monitor
+    4) Run subtest if necessary
+    5) Reset memory back to the original value
+
+    :param test: QEMU test object
+    :param params: Dictionary with the test parameters
+    :param env: Dictionary with test environment.
+    """
+
+    def _memory_check_after_sub_test():
+        """
+        Check memory status after subtest, the ballooned_mem is 0
+        """
+        output = balloon_test.memory_check("after subtest", 0)
+        return output
+
+    if params['os_type'] == 'windows':
+        balloon_test = BallooningTestPauseWin(test, params, env)
+    else:
+        balloon_test = BallooningTestPauseLinux(test, params, env)
+
+    min_sz, max_sz = balloon_test.get_memory_boundary()
+
+    for tag in params.objects('test_tags'):
+        vm = env.get_vm(params["main_vm"])
+        if vm.monitor.verify_status('paused'):
+            error_context.context("Running balloon %s test when"
+                                  " the guest in paused status" % tag,
+                                  logging.info)
+        else:
+            error_context.context("Running balloon %s test after"
+                                  " the guest turned to running status" % tag,
+                                  logging.info)
+        params_tag = params.object_params(tag)
+        balloon_type = params_tag['balloon_type']
+        if balloon_type == 'evict':
+            expect_mem = int(random.uniform(min_sz, balloon_test.old_mmem))
+        else:
+            expect_mem = int(random.uniform(balloon_test.old_mmem, max_sz))
+
+        balloon_test.balloon_memory(expect_mem)
+        ballooned_memory = abs(balloon_test.old_mmem - expect_mem)
+        mmem, gmem = balloon_test.memory_check("after %s memory" % tag,
```

Comment:
For the turn memory balloon in paused status, design to not check its guest memory.